controllers/clustercache/cluster_cache.go

```diff
@@ -169,34 +169,74 @@ type ClusterCache interface {
 // because there is no connection to the workload cluster.
 var ErrClusterNotConnected = errors.New("connection to the workload cluster is down")
 
-// Watcher is a scoped-down interface from Controller that only has the Watch func.
+// Watcher is an interface that can start a Watch.
 type Watcher interface {
-	// Watch watches the provided Source.
-	Watch(src source.Source) error
+	Name() string
+	Object() client.Object
+	Watch(cache cache.Cache) error
+}
+
+// SourceWatcher is a scoped-down interface from Controller that only has the Watch func.
+type SourceWatcher[request comparable] interface {
+	Watch(src source.TypedSource[request]) error
 }
 
 // WatchInput specifies the parameters used to establish a new watch for a workload cluster.
-// A source.Kind source (configured with Kind, EventHandler and Predicates) will be added to the Watcher.
-// To watch for events, the source.Kind will create an informer on the Cache that we have created and cached
+// A source.TypedKind source (configured with Kind, TypedEventHandler and Predicates) will be added to the Watcher.
+// To watch for events, the source.TypedKind will create an informer on the Cache that we have created and cached
+// for the given Cluster.
+type WatchInput = TypedWatchInput[client.Object, ctrl.Request]
+
+// TypedWatchInput specifies the parameters used to establish a new watch for a workload cluster.
+// A source.TypedKind source (configured with Kind, TypedEventHandler and Predicates) will be added to the Watcher.
+// To watch for events, the source.TypedKind will create an informer on the Cache that we have created and cached
 // for the given Cluster.
-type WatchInput struct {
+type TypedWatchInput[object client.Object, request comparable] struct {
```

Comment:
```suggestion
type TypedWatcherOptions[object client.Object, request comparable] struct {
```

+ WatchInput => WatcherOptions

I think this makes more sense now. We have a lot of other options structs here + WatchInput is not the input of a function called Watch anymore